icu4c/source/i18n/messageformat2.cpp

```diff
@@ -0,0 +1,1148 @@
+    switch (source[index]) {
+    case BACKSLASH:
+    case LEFT_CURLY_BRACE:
+    case RIGHT_CURLY_BRACE: {
+        index++; // Consume the character
+        break;
+    }
+    default: {
+        ERROR(parseError, errorCode, index);
+        break;
+    }
+    }
+    // Guarantee postcondition
+    CHECK_BOUNDS(source, index, parseError, errorCode);
+}
+
+/*
+  Consume a non-empty sequence of text characters and escaped text characters,
+  matching the `text` nonterminal in the grammar
+*/
+static void parseText(const UnicodeString &source,
+                      uint32_t &index,
+                      UParseError &parseError,
+                      UErrorCode &errorCode) {
+    CHECK_ERROR(errorCode)
+    U_ASSERT(inBounds(source, index));
+    bool empty = true;
+
+    while (true) {
+        if (source[index] == BACKSLASH) {
+            parseTextEscape(source, index, parseError, errorCode);
+        } else if (isTextChar(source[index])) {
+            index++;
+            // Restore precondition
+            CHECK_BOUNDS(source, index, parseError, errorCode);
+        } else {
+            break;
+        }
+        empty = false;
+    }
+
+    if (empty) {
+        // text must be non-empty
+        ERROR(parseError, errorCode, index);
+    }
+}
+
+/*
+  Consume an `nmtoken`, `literal`, or the string "*", matching
+  the `key` nonterminal in the grammar
+*/
+static void parseKey(const UnicodeString &source,
+                     uint32_t &index,
+                     UParseError &parseError,
+                     UErrorCode &errorCode) {
+    CHECK_ERROR(errorCode);
+    U_ASSERT(inBounds(source, index));
+
+    // Literal | nmtoken | '*'
+    switch (source[index]) {
+    case PIPE: {
+        parseLiteral(source, index, parseError, errorCode);
+        break;
+    }
+    case ASTERISK: {
+        index++;
+        // Guarantee postcondition
+        CHECK_BOUNDS(source, index, parseError, errorCode);
+        break;
+    }
+    default: {
+        // nmtoken
+        parseNmtoken(source, index, parseError, errorCode);
+        break;
+    }
+    }
+}
+
+/*
+  Consume a non-empty sequence of `key`s separated by whitespace
+*/
+static void parseNonEmptyKeys(const UnicodeString &source,
+                              uint32_t &index,
+                              UParseError &parseError,
+                              UErrorCode &errorCode) {
+    CHECK_ERROR(errorCode);
+    U_ASSERT(inBounds(source, index));
+
+    // The key sequence must begin with whitespace
+    parseRequiredWhitespace(source, index, parseError, errorCode);
+    // Restore precondition
+    CHECK_BOUNDS(source, index, parseError, errorCode);
+    parseKey(source, index, parseError, errorCode);
+
+    while (isWhitespace(source[index])) {
+        // Treat whitespace as optional here since the sequence doesn't
+        // need to end with whitespace
+        parseOptionalWhitespace(source, index, parseError, errorCode);
```

Comment:
It doesn't validate `when |foo|bar`, because `parseKey()` exits when it sees a non-whitespace character immediately after a key, so in this case it won't validate because `bar` is not a pattern.

I'm adding this and a few variations on it as tests.

Reply:
Added the test that shows that this works: https://github.com/unicode-org/icu/blob/b1490dcb68d8707d368c0ea630e3a687ce465be7/icu4c/source/test/intltest/messageformat2test.cpp#L422